src/core/ddsc/tests/config.c

```diff
@@ -118,3 +118,228 @@ CU_Test(ddsc_config, incorrect_config, .init = ddsrt_init, .fini = ddsrt_fini) {
+    };
+
+    dds_entity_t participant;
+
+    /* Set up the trace sinks to detect the config parsing. */
+    dds_set_log_mask(DDS_LC_FATAL|DDS_LC_ERROR|DDS_LC_WARNING|DDS_LC_CONFIG);
+    dds_set_log_sink(&logger, (void*)log_expected);
+    dds_set_trace_sink(&logger, (void*)log_expected);
+
+    /* Create participant with an empty security element. */
+    found = 0;
+    ddsrt_setenv(URI_VARIABLE, "<DDSSecurity/>");
+    participant = dds_create_participant(DDS_DOMAIN_DEFAULT, NULL, NULL);
+    ddsrt_setenv(URI_VARIABLE, "");
+    CU_ASSERT_FATAL(participant < 0);
+
+    /* All traces should have been provided. */
+#ifndef DDSI_INCLUDE_SECURITY
+    CU_ASSERT_FATAL(found == 0x1);
+#else
+    CU_ASSERT_FATAL(found == 0x7);
+#endif
+}
+
+CU_Test(ddsc_config, security_missing, .init = ddsrt_init, .fini = ddsrt_fini) {
+
+    /* Expected traces when creating participant with the security elements. */
+    const char *log_expected[] = {
+#ifndef DDSI_INCLUDE_SECURITY
+      "config: //CycloneDDS/Domain: DDSSecurity: unknown element*",
+#else
+      "config: //CycloneDDS/Domain/DDSSecurity/Authentication/IdentityCertificate/#text: element missing in configuration*",
+      "config: //CycloneDDS/Domain/DDSSecurity/Authentication/IdentityCA/#text: element missing in configuration*",
+      "config: //CycloneDDS/Domain/DDSSecurity/Authentication/PrivateKey/#text: element missing in configuration*",
+#endif
+      NULL
+    };
+
+    /* IdentityCertificate, IdentityCA and PrivateKey values or elements are missing. */
+    const char *sec_config =
+      "<Tracing><Verbosity>finest</></>"
+      "<DDSSecurity>"
+        "<Authentication>"
+          "<Library path=\"dds_security_auth\" initFunction=\"init_authentication\" finalizeFunction=\"finalize_authentication\" />"
+          "<IdentityCertificate></IdentityCertificate>"
+          "<PrivateKey></PrivateKey>"
+          "<Password>testtext_Password_testtext</Password>"
+        "</Authentication>"
+          "<Cryptographic>"
+            "<Library path=\"dds_security_crypto\" initFunction=\"init_crypto\" finalizeFunction=\"finalize_crypto\"/>"
+          "</Cryptographic>"
+        "<AccessControl>"
+          "<Library path=\"dds_security_ac\" initFunction=\"init_ac\" finalizeFunction=\"finalize_ac\"/>"
+          "<Governance>file:Governance.p7s</Governance>"
+          "<PermissionsCA>file:Permissions_CA.pem</PermissionsCA>"
+          "<Permissions>file:Permissions.p7s</Permissions>"
+        "</AccessControl>"
+      "</DDSSecurity>";
+
+
+    dds_entity_t participant;
+
+    ddsrt_setenv(URI_VARIABLE, sec_config);
+
+    /* Set up the trace sinks to detect the config parsing. */
+    dds_set_log_mask(DDS_LC_FATAL|DDS_LC_ERROR|DDS_LC_WARNING|DDS_LC_CONFIG);
+    dds_set_log_sink(&logger, (void*)log_expected);
+    dds_set_trace_sink(&logger, (void*)log_expected);
+
+    /* Create participant with an empty security element. */
+    found = 0;
+    ddsrt_setenv(URI_VARIABLE, sec_config);
```

Comment:
How can it work without `<CycloneDDS><Domain>` parent elements?

Reply:
I don't know. It's a configuration feature that 'just works'. I scanned the documentation, but could not find an explanation of the usage. Maybe @eboasson can explain a bit more about how the newer features of the configuration works?

Reply:
@martinbremmer yes, that'd be worth documenting ... I suppose the existing documentation is limited to the PR descriptions ...

Anyway: yes, if an XML fragment from the environment doesn't start with ``<CycloneDDS>``, it will insert it; and if it then doesn't start with ``<Domain>`` it will insert ``<Domain id="any">``. In this mode, it will also accept ``</>`` as a universal closing tag and close any open elements on end-of-input.

It doesn't do any of that trickery when from a file.